digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/lane/LaneService.scala

```diff
@@ -224,68 +289,110 @@ trait LaneOperations {
         }
 
         // To remove the lanes with wrong SideCode
         val toRemove = lanesToProcess.filterNot(_.id == 0L)
                                     .filter { lane =>
                                       val isLaneCodeTowards = lane.laneCode >= MainLane.towardsDirection && lane.laneCode <= FourthRightAdditional.towardsDirection
+                                      val isSideCodeTowardsOk = lane.sideCode != mainLane11SideCode.value && lane.sideCode != SideCode.BothDirections.value
+
                                       val isLaneCodeAgainst = lane.laneCode >= MainLane.againstDirection && lane.laneCode <= FourthRightAdditional.againstDirection
+                                      val isSideCodeAgainstOk = lane.sideCode != mainLane21SideCode.value && lane.sideCode != SideCode.BothDirections.value
 
-                                      (isLaneCodeTowards && lane.sideCode != mainLane11SideCode.value) ||
-                                          (isLaneCodeAgainst && lane.sideCode != mainLane21SideCode.value)
+                                      (isLaneCodeTowards && isSideCodeTowardsOk ) || (isLaneCodeAgainst && isSideCodeAgainstOk)
                                      }
                                     .map(_.id)
 
         val lanesToAdd = (lanes ++ toAdd).filterNot(lane => toRemove.contains(lane.id))
 
-        (lanesToAdd, changeSet.copy(generatedPersistedLanes = changeSet.generatedPersistedLanes ++ lanesToAdd, expiredLaneIds = changeSet.expiredLaneIds ++ toRemove))
+        val newChangeSet = changeSet.copy(generatedPersistedLanes = changeSet.generatedPersistedLanes ++ lanesToAdd,
+                                    expiredLaneIds = changeSet.expiredLaneIds ++ toRemove,
+                                    adjustedSideCodes = changeSet.adjustedSideCodes ++ needAdjustSideCode )
+
+        (lanesToAdd, newChangeSet)
 
       case TrafficDirection.TowardsDigitizing =>
 
-        val toAdd = if (!lanesToProcess.exists(lane => lane.laneCode == mainLaneDirectionOK && lane.sideCode == SideCode.BothDirections.value))
+
+
+        val toAdd = if (!lanesToProcess.exists(lane => lane.laneCode == mainLaneDirectionOK))
                       Seq(createPersistedLane(mainLaneDirectionOK, SideCode.BothDirections.value, baseLane.municipalityCode, baseProps))
                     else
                       Seq()
 
+        val (laneCodeStart, laneCodeEnd) = getLaneNumbersRange(mainLaneDirectionOK)
+        val needAdjustSideCode = getLaneCodesForSideCodeAdjust(lanesToProcess, laneCodeStart, laneCodeEnd, SideCode.BothDirections)
+
         val toRemove = lanesToProcess.filterNot(_.id == 0L)
                                     .filter(lane => (lane.laneCode >= mainLaneDirectionRemove._1 && lane.laneCode <= mainLaneDirectionRemove._2) ||
                                               lane.sideCode != SideCode.BothDirections.value)
                                     .map(_.id)
```

Comment:
Let's avoid to perform two filters on the same object, I think will be better if we do something like:
```Scala
        val toRemove = lanesToProcess.filter(lane =>
          lane.id != 0L && ((lane.laneCode >= mainLaneDirectionRemove._1 && lane.laneCode <= mainLaneDirectionRemove._2) || lane.sideCode != SideCode.BothDirections.value))
          .map(_.id)

Reply:
Done